sensor_msgs/include/sensor_msgs/point_cloud_conversion.hpp

```diff
@@ -0,0 +1,190 @@
+  * \param output the resultant message in the sensor_msgs::msg::PointCloud2 format
+  */
+static inline bool convertPointCloudToPointCloud2(
+  const sensor_msgs::msg::PointCloud & input,
+  sensor_msgs::msg::PointCloud2 & output)
+{
+  output.header = input.header;
+  output.width = input.points.size();
+  output.height = 1;
+  output.fields.resize(3 + input.channels.size());
+  // Convert x/y/z to fields
+  output.fields[0].name = "x"; output.fields[1].name = "y"; output.fields[2].name = "z";
+  int offset = 0;
+  // All offsets are *4, as all field data types are float32
+  for (size_t d = 0; d < output.fields.size(); ++d, offset += 4) {
+    output.fields[d].offset = offset;
+    output.fields[d].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    output.fields[d].count = 1;
+  }
+  output.point_step = offset;
+  output.row_step = output.point_step * output.width;
+  // Convert the remaining of the channels to fields
+  for (size_t d = 0; d < input.channels.size(); ++d) {
+    output.fields[3 + d].name = input.channels[d].name;
+  }
+  output.data.resize(input.points.size() * output.point_step);
+  output.is_bigendian = false;  // @todo ?
+  output.is_dense = false;
+
+  // Copy the data points
+  for (size_t cp = 0; cp < input.points.size(); ++cp) {
+    memcpy(&output.data[cp * output.point_step + output.fields[0].offset], &input.points[cp].x,
+      sizeof(float));
+    memcpy(&output.data[cp * output.point_step + output.fields[1].offset], &input.points[cp].y,
+      sizeof(float));
+    memcpy(&output.data[cp * output.point_step + output.fields[2].offset], &input.points[cp].z,
+      sizeof(float));
+    for (size_t d = 0; d < input.channels.size(); ++d) {
+      if (input.channels[d].values.size() == input.points.size()) {
+        memcpy(&output.data[cp * output.point_step + output.fields[3 + d].offset],
+          &input.channels[d].values[cp], sizeof(float));
+      }
+    }
+  }
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/** \brief Convert a sensor_msgs::msg::PointCloud2 message to a sensor_msgs::msg::PointCloud message.
+  * \param input the message in the sensor_msgs::msg::PointCloud2 format
+  * \param output the resultant message in the sensor_msgs::msg::PointCloud format
+  */
+static inline bool convertPointCloud2ToPointCloud(
+  const sensor_msgs::msg::PointCloud2 & input,
+  sensor_msgs::msg::PointCloud & output)
+{
+  output.header = input.header;
+  output.points.resize(input.width * input.height);
+  output.channels.resize(input.fields.size() - 3);
+  // Get the x/y/z field offsets
+  int x_idx = getPointCloud2FieldIndex(input, "x");
+  int y_idx = getPointCloud2FieldIndex(input, "y");
+  int z_idx = getPointCloud2FieldIndex(input, "z");
+  if (x_idx == -1 || y_idx == -1 || z_idx == -1) {
+    std::cerr << "x/y/z coordinates not found! Cannot convert to sensor_msgs::msg::PointCloud!" <<
```

Comment:
Can't say I love using `cerr` here; that can be randomly noisy, won't get picked up by logging, etc.  I'd suggest either returning silently (this is the only reason that this function can return false), or throwing an exception.

Reply:
Removed.